File: src/main/java/ru/romangr/catbot/executor/RateLimiter.java
package ru.romangr.catbot.executor;

import com.google.common.cache.CacheBuilder;
import com.google.common.cache.CacheLoader;
import com.google.common.cache.LoadingCache;
import java.time.Duration;
import java.util.Map;
import java.util.Objects;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.stream.Stream;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;
import ru.romangr.catbot.telegram.model.Chat;

@Slf4j
public class RateLimiter {

  private static final int CHAT_ACTIONS_PER_MINUTE_LIMIT = 20;

  private final LoadingCache<Integer, AtomicInteger> chats;
  private final Map<Integer, Boolean> chatsToSkip;

  public RateLimiter() {
    this(Duration.ofMinutes(1), Duration.ofMinutes(1));
  }

  RateLimiter(Duration limitPeriod, Duration banTimeout) {
    this.chats = CacheBuilder.newBuilder()
        .expireAfterWrite(limitPeriod)
        .maximumSize(1000)
        .build(CacheLoader.from(key -> new AtomicInteger()));

    this.chatsToSkip = CacheBuilder.newBuilder()
        .expireAfterAccess(banTimeout)
        .maximumSize(1000)
        .build(CacheLoader.<Integer, Boolean>from(key -> {
          throw new RuntimeException("Should not be used");
        }))
        .asMap();
  }

  @SneakyThrows
  public RateLimitResult check(Chat chat) {
    int actionsCount = chats.get(chat.getId()).incrementAndGet();
    if (chatsToSkip.getOrDefault(chat.getId(), false)) {
      return RateLimitResult.BANNED;
    }
    if (actionsCount > CHAT_ACTIONS_PER_MINUTE_LIMIT) {
      chatsToSkip.put(chat.getId(), true);
      log.warn("Chat {} with id {} has been banned because of too many actions",
          getChatName(chat), chat.getId());
      return RateLimitResult.MADE_BANNED;
    }
    return RateLimitResult.POSITIVE;
  }

  private String getChatName(Chat chat) {
    return Stream.of(chat.getFirstName(), chat.getLastName(), chat.getTitle(), chat.getUsername())
        .filter(Objects::nonNull)
        .findFirst()
        .map(name -> "'" + name + "'")
        .orElse("");
  }

}

File: src/main/java/ru/romangr/catbot/executor/RateLimiter.kt
package ru.romangr.catbot.executor

import com.google.common.cache.CacheBuilder
import com.google.common.cache.CacheLoader
import com.google.common.cache.LoadingCache
import org.slf4j.LoggerFactory
import ru.romangr.catbot.telegram.model.Chat
import java.time.Duration
import java.util.Objects
import java.util.concurrent.ExecutionException
import java.util.concurrent.atomic.AtomicInteger
import java.util.stream.Stream

class RateLimiter internal constructor(limitPeriod: Duration, banTimeout: Duration) {
    private val chats: LoadingCache<Int, AtomicInteger> = CacheBuilder.newBuilder()
            .expireAfterWrite(limitPeriod)
            .maximumSize(1000)
            .build(CacheLoader.from { _: Int? -> AtomicInteger() })

    private val chatsToSkip: MutableMap<Int, Boolean> = CacheBuilder.newBuilder()
            .expireAfterAccess(banTimeout)
            .maximumSize(1000)
            .build(CacheLoader.from<Int, Boolean> {
                throw RuntimeException("Should not be used")
            })
            .asMap()

    constructor() : this(Duration.ofMinutes(1), Duration.ofMinutes(1))

    fun check(chat: Chat): RateLimitResult {
        val actionsCount: Int
        try {
            actionsCount = chats[chat.id].incrementAndGet()
        } catch (e: ExecutionException) {
            log.warn("Error incrementing rate limiter counter", e.cause)
            return RateLimitResult.BANNED
        }
        if (chatsToSkip.getOrDefault(chat.id, false)) {
            return RateLimitResult.BANNED
        }
        if (actionsCount > CHAT_ACTIONS_PER_MINUTE_LIMIT) {
            chatsToSkip[chat.id] = true
            log.warn("Chat {} with id {} has been banned because of too many actions",
                    getChatName(chat), chat.id)
            return RateLimitResult.MADE_BANNED
        }
        return RateLimitResult.POSITIVE
    }

    private fun getChatName(chat: Chat): String {
        return Stream.of(chat.firstName, chat.lastName, chat.title, chat.username)
                .filter { obj: String? -> Objects.nonNull(obj) }
                .findFirst()
                .map { name: String? -> "'$name'" }
                .orElse("")
    }

    companion object {
        private val log = LoggerFactory.getLogger(RateLimiter::class.java)
        private const val CHAT_ACTIONS_PER_MINUTE_LIMIT = 20
    }
}
